File: data/src/main/java/home/oleg/placesnearme/data/repositories/UserLocationRepositoryImpl.java
package home.oleg.placesnearme.data.repositories;

import home.oleg.placenearme.models.UserLocation;
import home.oleg.placenearme.repositories.UserLocationRepository;
import home.oleg.placesnearme.data.mapper.LocationMapper;
import home.oleg.placesnearme.data.provider.CachedLocationsStore;
import home.oleg.placesnearme.data.provider.ReactiveLocationStore;
import io.reactivex.Single;

/**
 * Created by Oleg Sheliakin on 17.09.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
public class UserLocationRepositoryImpl implements UserLocationRepository {

    private final ReactiveLocationStore reactiveLocationStore;
    private final CachedLocationsStore cachedLocationsStore;

    public UserLocationRepositoryImpl(ReactiveLocationStore reactiveLocationStore, CachedLocationsStore cachedLocationsStore) {
        this.reactiveLocationStore = reactiveLocationStore;
        this.cachedLocationsStore = cachedLocationsStore;
    }

    @Override
    public Single<UserLocation> getLocation() {
        return reactiveLocationStore
                .getLastLocation()
                .doOnSuccess(cachedLocationsStore::save)
                .onErrorResumeNext(cachedLocationsStore.getLastLocation())
                .map(LocationMapper::map);
    }

}

File: core-data/src/main/java/home/oleg/placesnearme/data/repositories/UserLocationRepositoryImpl.kt
package home.oleg.placesnearme.data.repositories

import home.oleg.placenearme.models.UserLocation
import home.oleg.placenearme.repositories.UserLocationRepository
import home.oleg.placesnearme.data.mapper.LocationMapper
import home.oleg.placesnearme.data.location.CachedLocationsStore
import home.oleg.placesnearme.data.location.ReactiveLocationStore
import io.reactivex.Single
import javax.inject.Inject

/**
 * Created by Oleg Sheliakin on 17.09.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
class UserLocationRepositoryImpl @Inject constructor(
        private val reactiveLocationStore: ReactiveLocationStore,
        private val cachedLocationsStore: CachedLocationsStore) : UserLocationRepository {

    override val location: Single<UserLocation>
        get() = reactiveLocationStore
                .lastLocation
                .doOnSuccess { cachedLocationsStore.save(it) }
                .onErrorResumeNext(cachedLocationsStore.lastLocation)
                .map { LocationMapper.map(it) }

}
